_posts/2021-07-26-hhvm-4.120.markdown

```diff
@@ -0,0 +1,66 @@
+---
+title: "HHVM 4.120"
+layout: post
+author: fred
+category: blog
+---
+
+HHVM 4.120 is released! This release marks the end of support for 4.113 and
+4.114; HHVM 4.115&ndash;4.118 remain supported, as do the 4.80 and 4.102 LTS
+releases. HHVM 4.119 was skipped due to personal circumstances of the team.
+
+
+# Highlights
+
+- autocomplete can now provide documentation for suggestions
+- autocomplete now understands built-in primitive types like `string`
+- `stream_select()` and `pagelet_server_task_result()` will no longer be
+  inlined, so will appear correctly in profilers like Xenon.
+- added `HH\\Lib\\Legacy_FIXME` namespace, containing several functions emulating
+  PHP behavior that we intend to remove from the core language. New code should
+  never use these functions, and projects should aim to remove them over time.
+  Currently, these include:
+  - `Legacy_FIXME\\increment(mixed)` and `decrement(mixed)`: reimplements the
+    current behavior of `++` and `--` on `null` and `string` inputs.
```

Comment:
The pre/post increment/decrement operators change the value of the current variable. These functions do not take their argument with inout, so they do not change the current variable. You need to assign the return value of these functions to the current variable.

I am being very pedantic here, sorry 🙂.

Reply:
These functions don't feel like hsl functions btw. The `HH\FIXME` namespace (where `HH\FIXME\UNSAFE_CAST` lives) would have been more fitting. It is easier to write tests in the hsl directory, because of MiniTest, so you win some, you lose some.

Reply:
Yeah; I'm seeing the primary difference as:
- `HH\FIXME` is for type errors
- `HH\Lib\Legacy_FIXME` is for runtime behaviors

We could use either namespace in `hphp/hsl/`; I'm not certain we've got the right place, but it seems good enough, especially with them being intended as a legacy migration aid only